community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java

```diff
@@ -517,23 +517,27 @@ private void commit() throws TransactionFailureException
         try ( CommitEvent commitEvent = transactionEvent.beginCommitEvent() )
         {
             // Trigger transaction "before" hooks.
-            if ( hasDataChanges() )
+            if ( hasTxStateWithChanges() )
             {
-                try
+                if ( txState.hasDataChanges() )
                 {
-                    if ( (hooksState = hooks.beforeCommit( txState, this, storeLayer )) != null && hooksState.failed() )
+                    try
                     {
-                        throw new TransactionFailureException( Status.Transaction.HookFailed, hooksState.failure(),
-                                "" );
+                        if ( (hooksState = hooks.beforeCommit( txState, this, storeLayer )) != null && hooksState.failed() )
+                        {
+                            throw new TransactionFailureException( Status.Transaction.HookFailed, hooksState.failure(),
+                                    "" );
+                        }
+                    }
+                    finally
+                    {
+                        beforeHookInvoked = true;
                     }
                 }
-                finally
-                {
-                    beforeHookInvoked = true;
-                }
-            }
 
-            prepareRecordChangesFromTransactionState();
+                context.init( locks );
+                prepareRecordChangesFromTransactionState();
```

Comment:
we need it since converting some logical tx state changes to record changes requires additional locks to be acquired. Rollback doesn't.